File: PeekingIterator.js

/**
 * @param {Iterator} iterator
 */
var PeekingIterator = function (iterator) {
    this.iterator = iterator;
    this.nextLoaded = iterator.hasNext() ? iterator.next() : null;
};

/**
 * @return {number}
 */
PeekingIterator.prototype.peek = function () {
    if (!this.hasNext()) {
        throw "No Such Element Exception.";
    }
    return this.nextLoaded;
};

/**
 * @return {number}
 */
PeekingIterator.prototype.next = function () {
    if (!this.hasNext()) {
        throw "No Such Element Exception.";
    }

    let temp = this.nextLoaded;
    this.nextLoaded = null;

    if (this.iterator.hasNext()) {
        this.nextLoaded = this.iterator.next();
    }

    return temp;
};

/**
 * @return {boolean}
 */
PeekingIterator.prototype.hasNext = function () {
    return this.nextLoaded !== null;
};

/**
 *  This is the Iterator's API interface. 
 *  It is implemented internally by leetcode.com.
 *  It should not be included in the code, when running it on the website.
 *  
 * function Iterator() {
 *    @return {number}
 *    this.next = function() { // return the next number of the iterator
 *       ...
 *    }; 
 *
 *    @return {boolean}
 *    this.hasNext = function() { // return true if it still has numbers
 *       ...
 *    };
 * };
 */

File: PeekingIterator.cpp

#include<vector>
using namespace std;

class PeekingIterator : public Iterator {
public:

    unique_ptr<int> nextLoaded;

    PeekingIterator(const vector<int>& nums) : Iterator(nums) {
        if (Iterator::hasNext()) {
            this->nextLoaded = make_unique<int>(Iterator::next());
        }
    }

    int peek() {
        if (!hasNext()) {
            throw ("No Such Element Exception.");
        }
        return *nextLoaded;
    }

    int next() {
        if (!hasNext()) {
            throw ("No Such Element Exception.");
        }

        int temp = *nextLoaded;
        nextLoaded = nullptr;

        if (Iterator::hasNext()) {
            nextLoaded = make_unique<int>(Iterator::next());
        }
        return temp;
    }

    bool hasNext() const {
        return nextLoaded != nullptr;
    }
};

/*
 *  This is the Iterator's API interface. 
 *  It is implemented internally by leetcode.com.
 *  It should not be included in the code, when running it on the website.
 *
 *  class Iterator {
 *	struct Data;
 * 	Data* data;
 *  public:
 *	Iterator(const vector<int>& nums);
 * 	Iterator(const Iterator& iter);
 *
 * 	// Returns the next element in the iteration.
 *	int next();
 *
 *	// Returns true if the iteration has more elements.
 *	bool hasNext() const;
 *  };
 */
